common/malloc_utils/src/glibc.rs

```diff
@@ -0,0 +1,194 @@
+const M_MMAP_THRESHOLD: c_int = -4;
+const M_ARENA_MAX: c_int = -8;
+
+/// Environment variables used to configure malloc.
+///
+/// Source:
+///
+/// https://man7.org/linux/man-pages/man3/mallopt.3.html
+const ENV_VAR_ARENA_MAX: &str = "MALLOC_ARENA_MAX";
+const ENV_VAR_MMAP_THRESHOLD: &str = "MALLOC_MMAP_THRESHOLD_";
+
+#[allow(dead_code)]
+enum ArenaMaxSetting {
+    /// Do not set any value for MALLOC_ARENA_MAX, leave it as default.
+    DoNotSet,
+    /// Set a fixed value.
+    Fixed(c_int),
+    /// Read the number of CPUs at runtime and use that value.
+    NumCpus,
+}
+
+// Metrics for the malloc. For more information, see:
+//
+// https://man7.org/linux/man-pages/man3/mallinfo.3.html
+lazy_static! {
+    pub static ref MALLINFO_ARENA: lighthouse_metrics::Result<IntGauge> = try_create_int_gauge(
+        "mallinfo_arena",
+        "The total amount of memory allocated by means other than mmap(2). \
+            This figure includes both in-use blocks and blocks on the free list.",
+    );
+    pub static ref MALLINFO_ORDBLKS: lighthouse_metrics::Result<IntGauge> = try_create_int_gauge(
+        "mallinfo_ordblks",
+        "The number of ordinary (i.e., non-fastbin) free blocks.",
+    );
+    pub static ref MALLINFO_SMBLKS: lighthouse_metrics::Result<IntGauge> =
+        try_create_int_gauge("mallinfo_smblks", "The number of fastbin free blocks.",);
+    pub static ref MALLINFO_HBLKS: lighthouse_metrics::Result<IntGauge> = try_create_int_gauge(
+        "mallinfo_hblks",
+        "The number of blocks currently allocated using mmap.",
+    );
+    pub static ref MALLINFO_HBLKHD: lighthouse_metrics::Result<IntGauge> = try_create_int_gauge(
+        "mallinfo_hblkhd",
+        "The number of bytes in blocks currently allocated using mmap.",
+    );
+    pub static ref MALLINFO_FSMBLKS: lighthouse_metrics::Result<IntGauge> = try_create_int_gauge(
+        "mallinfo_fsmblks",
+        "The total number of bytes in fastbin free blocks.",
+    );
+    pub static ref MALLINFO_UORDBLKS: lighthouse_metrics::Result<IntGauge> = try_create_int_gauge(
+        "mallinfo_uordblks",
+        "The total number of bytes used by in-use allocations.",
+    );
+    pub static ref MALLINFO_FORDBLKS: lighthouse_metrics::Result<IntGauge> = try_create_int_gauge(
+        "mallinfo_fordblks",
+        "The total number of bytes in free blocks.",
+    );
+    pub static ref MALLINFO_KEEPCOST: lighthouse_metrics::Result<IntGauge> = try_create_int_gauge(
+        "mallinfo_keepcost",
+        "The total amount of releasable free space at the top of the heap..",
+    );
+}
+
+/// Calls `mallinfo` and updates Prometheus metrics with the results.
+pub fn scrape_mallinfo_metrics() {
+    // The docs for this function say it is thread-unsafe since it may return inconsistent results.
+    // Since these are just metrics it's not a concern to us if they're sometimes inconsistent.
```

Comment:
I'm not sure if this interpretation is right, and I'm a bit worried we might introduce memory unsafety or corrupt malloc's internal datastructures.

The `mallinfo` function is marked [here](https://man7.org/linux/man-pages/man3/mallinfo.3.html#ATTRIBUTES) as:

```
MT-Unsafe init const:mallopt
```

Interpreting this according to the [description of attributes](https://man7.org/linux/man-pages/man7/attributes.7.html):

* _**MT-Unsafe** functions are not safe to call in multithreaded programs._ This sounds like it's just generically unsafe, and could invoke undefined behaviour. [Other GNU libc docs](https://www.gnu.org/software/libc/manual/html_mono/libc.html#POSIX-Safety-Concepts) support this interpretation: _MT-Unsafe, AS-Unsafe, AC-Unsafe functions are not safe to call within the safety contexts described above. Calling them within such contexts invokes undefined behavior._
* **init**: _Functions marked with init as an MT-Unsafe feature perform MT-Unsafe initialization when they are first called._ This would imply that we could remove some unsafety by calling `mallinfo` in a single-threaded context on startup, but I imagine that would be quite clunky.
* **const:mallopt**: I think this is saying that it modifies data that `mallopt` modifies _without locking_, causing the potential for races/undefined behaviour if `mallopt` calls aren't synchronised (currently they aren't).

Given this I wonder if we shouldn't use `malloc_info` instead: https://man7.org/linux/man-pages/man3/malloc_info.3.html. It's thread-safe, but would require us to parse XML :star_struck: ...

I also vaguely recall you mentioning that the malloc metrics were incomplete, which could be explained by this para from `mallinfo`'s docs:

> Information is returned for only the main memory allocation area.
       Allocations in other arenas are excluded.  See malloc_stats(3)
       and malloc_info(3) for alternatives that include information
       about other arenas.

Finally, if we do keep `mallinfo`, we should at least use `mallinfo2` (`mallinfo` is deprecated).

Reply:
>  I'm a bit worried we might introduce memory unsafety or corrupt malloc's internal datastructures.

Good call, I'll guard these calls with a `Mutex`. Thanks.

> Given this I wonder if we shouldn't use malloc_info instead: https://man7.org/linux/man-pages/man3/malloc_info.3.html

I spent a fair bit of time trying to get this to work but I just couldn't get it to return that XML string.

> Finally, if we do keep mallinfo, we should at least use mallinfo2 (mallinfo is deprecated).

The docs for this say *"The mallinfo2() function was added in glibc 2.33."*, but 2.33 was [released 2021-02-01](https://www.gnu.org/software/libc/). I was concerned this wouldn't exist for some of our users. I'll add a comment to describe this.

Reply:
Oh I didn't notice that about `mallinfo2`!

Happy with `mallinfo` + `Mutex` solution.